ecl/eclcc/WORKUNITS.rst

```diff
@@ -0,0 +1,862 @@
+for some roxie queries (e.g., in a production system) the execution workunits are transient.
+Whichever path is followed, the following walk-through details the main stages in executing a query, and the effect each
+of the query elements has.
+
+Workflow
+========
+When a workunit is ready to be run, the workflow controls the flow of execution.
+The workflow engine (common/workunit/workflow.cpp) is responsible for determining which workflow item should be
+executed next.  (The workflow for Thor and hThor jobs is coordinated by eclagent, while roxie includes the workflow
+engine in its process.  The eclscheduler also uses the workflow engine to process events and mark workflow items ready
+to be executed.)
+
+eclagent, or roxie calls the createProcess() function from the workunit dll to create an instance of the generated workflow
+helper class, and passes it to the workflow engine.  The workflow engine walks the workflow items to find any items
+which are are ready to be executed (have the state "reqd").  Dependencies on other workflow items are executed first,
+and if they succeed the workflow item is executed.  The example has the following workflow entries::
+
+ <Workflow>
+  <Item mode="normal"
+        state="null"
+        type="normal"
+        wfid="1"/>
+  <Item mode="normal"
+        state="reqd"
+        type="normal"
+        wfid="2">
+   <Dependency wfid="1"/>
+   <Schedule/>
+  </Item>
+ </Workflow>
+
+Item 2 has a state of "reqd", so it should be evaluated now.  Item 2 has a dependency on item 1, so that must be
+evaluated first.  This is acheived by calling MyEclProcess::perform() on the object that was previously created
+from the workunit dll, passing in wfid = 1.  That will execute the following code::
+
+    switch (wfid) {
+        case 1U:
+            if (!gctx->isResult("searchname",4294967295U)) {
+                ctx->setResultString("searchname",4294967295U,5U,"Smith");
+            }
+            break;
+        break;
+    }
+
+This checks if a value has been provided for the input parameter, and if not assigns a default value of "Smith".
+The function returns and control returns to the workflow engine.  With the dependencies for wfid 2 now satisfied,
+the generated code for that workflow id is now executed::
+
+    switch (wfid) {
+        case 2U: {
+            ctx->executeGraph("graph1",false,0,NULL);
+            ctx->setResultString(0,1U,5U,"Done!");
+        }
+        break;
+    }
+    
+Most of the work for this workflow item involves executing graph1 (by calling back into the eclagent/roxie).  However
+it also directly sets another result.  This is fairly typical - the code inside MyProcess::perform often combines
+evaluating scalar results and temporary expressions, executing graphs, and calling functions
+that cannot (currently) be called inside a graph (e.g., those involving superfile transactions).
```

Comment:
Superfiles haven't been mentioned before, a brief explanation should be required.